### ISSUE TITLE
Routing when driving on toll highway and "No toll road" option activated 

### ISSUE THREAD

**Issue by MarcMart** [osmandapp/Osmand#8085]
Ho OsmAnd developers,
first of all congrats for the great job you are doing, this app is awesome! Second, please apologize for my inexperience as it is the first thing I have ever done on GitHub, I have created an account just for you guys!

OSM user:  Mar Mar
Android Version: 3.5.5 (the issue has always existed, no matter the version)
Offline Maps Spain or France

**Issue name:** Routing when driving on toll highway and "No toll road" option activated 
**Issue Description:** 
The app makes no distinction between the highways you WISH to be routed on (incl. restrictions) and where you are REALLY located. The result is that if you happen to drive on a toll highway and chose to get off asap, activating the "No toll road" option, the app assumes you cannot possibly drive on one at the same time. As a consequence it will try to calculate a route from the nearest non-highway location. 
And here is were the problem starts:

Driving along, the nearest non-toll highway location will change constantly, forcing the app to recalculate one erroneous route after another (see screenshot example) . Drawbacks are that you are effectively not routed off the toll highway and that while you keep driving on a toll highway you are not provided with critical info you need to make a decision (e.g. what would be the distance/time until arrival IF you would be routed off the toll road) and can't compare both options. This becomes even worse if the destination is far away (several hundreds of km) and every recalculation takes a lot of time, with the result that the app won't even provide a new (erroneous) route as a new nearest non-toll location has been found before finishing the previous one.   

**Proposed solution:**
When driving on a toll highway and activating "No toll road" option, the app should find the nearest place to exit the toll road and continue routing on non-toll highways

Thanks for listening and best regards, Please feel free to get back to me in case of doubts. Any effort to resolve this issue is highly appreciated. 
Marcos Martinez

![Routing when driving on toll highway and _No toll road_ option activated](https://user-images.githubusercontent.com/58865443/70836573-d6c09500-1e00-11ea-89ac-16115139706b.png)

**Comment by OpenSourceForEver** [comment]
If you want a route without tolls, why are you driving on a toll highway?


If you select the option to avoid tolls it is logical that Osmand calculates a route without going through any toll road

**Comment by MarcMart** [comment]
Thanks for responding so quickly...
Your question seems logical at first glance. Not that it happens every day but there can be several situations and I will describe a couple, hoping not to bore you as the first needs more than a single sentence (there can be many others...):

1. The existence of toll highways relies on charging a fee and in turn give you the convenience of  arriving quicker at your destination. Nevertheless, while sometimes the travel time delta between a toll highway and the alternative route is huge, other times the difference is so little that it doesn't pay off. Now, especially when traveling longer distances both situation can happen as you can see below, heading in N-E direction towards Girona:

![Switch from toll to non-toll example](https://user-images.githubusercontent.com/58865443/70869830-fa184b00-1f8c-11ea-9deb-498d5648c3c8.png)

When you already know what to do, enabling navigation off the toll highway might just be slightly convenient but becomes priceless when you find yourself  (e.g. as a tourist) in an area you don't know, take the toll highway to be on the safe side and want to check what difference it would make to drive off at a given point of time. Right now in OsmAnd you simply never know until you actually have already driven off.

2. You're driving on a highway toll with very dense traffic and wonder what the alternative would be if you left it.

Hoping this helps to understand, best regards and with deepest respect for your work,
Marcos Martinez

**Comment by vshcherb** [comment]
This is an interesting situation that Avoid toll road completely avoids toll road. Probably we can apply similar approach as we have for 'avoid motorways'. 
You can download routing.xml and give it try yourself and share result. Changing routing.xml is supported for custom profiles https://github.com/osmandapp/OsmAnd-resources/blob/master/routing/routing.xml

**Comment by MarcMart** [comment]
Hi Victor, 

Terribly sorry but I too dumb to follow your instruction. I haven't been
able to download any file. Is there any other way or more detailed
instructions? Note I'm just a simple user and OSM contributor but I
haven'tused GitHub before. 

Happy new year in any case to you and the rest of the team! 

Marcos 

Am 24.12.2019 09:57, schrieb vshcherb:

> This is an interesting situation that Avoid toll road completely avoids toll road. Probably we can apply similar approach as we have for 'avoid motorways'.
> You can download routing.xml and give it try yourself and share result. Changing routing.xml is supported for custom profiles https://github.com/osmandapp/OsmAnd-resources/blob/master/routing/routing.xml 
> 
> --
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub [1], or unsubscribe [2].

 

Links:
------
[1]
https://github.com/osmandapp/Osmand/issues/8085?email_source=notifications&amp;email_token=AOBDOI5DW62WFDYLOEKZW5DQ2HFJDA5CNFSM4J2XBO22YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEHS2OWA#issuecomment-568698712
[2]

**Comment by vshcherb** [comment]
@MarcMart  no worries we're going to look ourselves.

**Comment by beatnickgr** [comment]
Same issue here :

BACKGROUND
I dislike to pay tolls so I have set to avoid toll roads.
But in 1 specific case, to avoid an unmaintained track on the mountain, I paid the toll and go through the tunnel, without changing the settings. 

BUG
After I paid the toll, osmand~ refused to see I was on the toll way already. It was "correcting" my location to the nearest mountain track, and recalculated my route from where I was NOT.

DIFFERENT COUNTRIES :
In Greece, the toll station is on the road and we pay by segment, so we want to stay on the segment as much as possible, and exit before the next toll station to avoid paying again.
In France and Spain, the toll station is at the exit and we pay by distance, so we want to take the 1st exit.

EXPECTED BEHAVIOUR 
Avoid toll stations, not toll roads (and rename the feature).
If already on a toll road, route through an exit without a station. (Will work in countries like Greece etc)
If there's no way to avoid toll stations on exits, route through the nearest one. (France, Spain etc)

**Comment by scaidermern** [comment]
> EXPECTED BEHAVIOUR
> Avoid toll stations, not toll roads (and rename the feature).

This is definitely an interesting approach. However I'm afraid this won't work for all countries since some countries don't have toll stations. Examples are motorways with tolls in Austria or Czechia. So I guess OsmAnd has to know for which countries to avoid toll stations and for which to avoid toll roads.

**Comment by MarcMart** [comment]
I agree, including the toll stations as parameter makes everything complicated and is actually not needed. If highways are correctly tagged (and they usually are) the logic "get me away from the toll section asap" does the trick, also in Greece. Let me describe what should happen in detail: You have just passed the toll station but are still on a toll tagged highway. You select the "avoid toll" option and now Osmand should find the nearest highway that is toll free - which of course should always be the first toll free section of the road you are already driving on. From this point it recalculates and should get you off teh highway at the last exit before the next toll section.

**Comment by beatnickgr** [comment]
The issue is not solved on this segment https://www.openstreetmap.org/way/27417638
Please reopen the ticket.

**Comment by MarcMart** [comment]
ok, I closed it by mistake

**Comment by eliwavv** [comment]
Has there been any updates on this issue?

**Comment by septatrix** [comment]
Seems to also occur when avoiding unpaved roads while on one
